libp2p/multiaddress.ss

```diff
@@ -0,0 +1,46 @@
+;;; -*- Gerbil -*-
+;;; multiaddress implementation
+
+(import :std/io/interface
+        :std/sugar
+        :std/misc/alist
+        :std/misc/list
+        :std/format
+        :std/iter
+        "utils")
+
+(export #t)
+
+;; A macro to define a predicate against the protocols in a multiaddress
+(defrule (def-multipattern name . protocols)
+  (def (name ma)
+    (andmap (lambda (protocol) (get-protocol-value ma protocol)) 'protocols)))
+
+(defstruct multi-address (protocols str)
+  constructor: init!)
+
+(defmethod {init! multi-address}
```

Comment:
it is also very useful to initialize from binary input (a u8vector), this is very common pattern that we need to support.

Reply:
My general recommendation would be to store the binary representation (as a u8vector) and a parsed version that is a alist/plist of protocols and protocol values. 
We can also cache the string rep for printing...

Actually I think it is better to have the binary rep, string rep, and list rep. The primary one is the list rep, that's what we should work with. And we can initialize either from string or binary rep, and generate the other one when needed (say to print or to send over the network).